design/resource-management.md

```diff
@@ -0,0 +1,329 @@
+
+root     23992     1  0 22:13 ?        00:00:00 /opt/kata/bin/containerd-shim-kata-v2 -namespace k8s.io -address /run/containerd/containerd.sock -publish-binary /usr/local/bin/containerd -id 9d17d1c5d075ca42dfefb58ef7b82c8b1b234cc128ebf9332b902b866c0ebed7 -debug
+root@kata-k8s-containerd:/sys/fs/cgroup/cpu,cpuacct# grep -ir 23992
+system.slice/containerd.service/cgroup.procs:23992
+system.slice/containerd.service/tasks:23992
+```
+
+qemu:
+```bash
+root@kata-k8s-containerd:/sys/fs/cgroup/cpu,cpuacct# ps -ae | grep qemu
+24007 ?        00:37:26 qemu-system-x86
+
+root@kata-k8s-containerd:/sys/fs/cgroup/cpu,cpuacct# grep -ir 24007
+kubepods/podf277e232-5ca6-11e9-b514-000d3a6d0876/9d17d1c5d075ca42dfefb58ef7b82c8b1b234cc128ebf9332b902b866c0ebed7/cgroup.procs:24007
+system.slice/containerd.service/cgroup.procs:24007
+system.slice/containerd.service/tasks:24007
+```
+
+vhost:
+```bash
+root@kata-k8s-containerd:/sys/fs/cgroup/cpu,cpuacct# ps -aef | grep vhost | grep -v qemu | grep -v grep
+root     24010     2  0 22:13 ?        00:00:00 [vhost-24007]
+
+root@kata-k8s-containerd:/sys/fs/cgroup/cpu,cpuacct# grep -ir 24010
+system.slice/containerd.service/cgroup.procs:24010
+system.slice/containerd.service/tasks:24010
+```
+
+#### In Kubernetes  + CRI-O (v1 shim)
+
+CRI-O is very similar the containerd, except for where the non-constrained processes end up. Instead
+of being called by CRIO directly, kata-runtime is called from a process `conmon`, which is located
+in a cgroup under the pod-cgroup. As expected based on prior exapmles, cgroups are ceated for each
+container, and the QEMU vCPU threads are placed within the pause containers cgroup.
+
+```bash
+pod1cc61d33-5ca1-11e9-90bc-525400cfa589/crio-1b05886a39901ef3a7555796d38dcfaafd8fda929aef223ea576324a4949f9ef/tasks
+{qemu-system-x86}(18061)
+kata-shim(18207)─┬─{kata-shim}(18213)
+                 ├─{kata-shim}(18215)
+                 ├─{kata-shim}(18216)
+                 ├─{kata-shim}(18217)
+                 ├─{kata-shim}(18218)
+                 ├─{kata-shim}(18219)
+                 ├─{kata-shim}(18220)
+                 ├─{kata-shim}(18221)
+                 ├─{kata-shim}(18223)
+                 ├─{kata-shim}(18224)
+                 └─{kata-shim}(18226)
+{kata-shim}(18213)
+{kata-shim}(18215)
+{kata-shim}(18216)
+{kata-shim}(18217)
+{kata-shim}(18218)
+{kata-shim}(18219)
+{kata-shim}(18220)
+{kata-shim}(18221)
+{kata-shim}(18223)
+{kata-shim}(18224)
+{kata-shim}(18226)
+{qemu-system-x86}(18280)
+{qemu-system-x86}(18281)
+{qemu-system-x86}(18368)
+{qemu-system-x86}(18369)
+{qemu-system-x86}(18370)
+```
+
+The QEMU, vhost, proxy and shim threads, however, are placed under the caller's cgroup, which in this case
+is `conmon`, which is a peer of the container cgroups we created. So, the good news is that QEMU, vhost,
+etc, are constrained within the pod's cgroup. The bad news is these will be constrained based on the values
+associated with conmon:
```

Comment:
This is also wrong as conmon should be constrained with the absolute minmum resources it can get away with. Typically this is set to 2 (i.e. so that it does not impact the container scheduling).